src/lib.rs

```diff
@@ -172,34 +169,93 @@ impl<R: Read> Archive<R> {
                                           "empty file name in tarball"))
             }
             let is_directory = bytes[bytes.len() - 1] == b'/';
+
+            // Notes regarding bsdtar 2.8.3 / libarchive 2.8.3:
+            // * Leading '/'s are trimmed. For example, `///test` is treated as
+            //   `test`.
+            // * If the filename contains '..', then the file is skipped when
+            //   extracting the tarball.
+            // * '//' within a filename is effectively skipped. An error is
+            //   logged, but otherwise the effect is as if any two or more
+            //   adjacent '/'s within the filename were consolidated into one
+            //   '/'.
+
             let mut dst = into.to_path_buf();
+            let mut contains_dot_dot = false;
+            let mut seen_non_empty_part = false;
             for part in bytes.split(|x| *x == b'/') {
+                if part == b".." {
+                    contains_dot_dot = true;
+                    break;
+                }
+                // If empty, skip this part.
+                // This effectively results in trimming leading '/'s as well as
+                // merging adjacent '/'s.
+                if part.len() == 0 {
+                    continue;
+                }
+                seen_non_empty_part = true;
                 push(&mut dst, part);
             }
+            // If any part of the filename is '..', then skip over unpacking the
+            // file to prevent directory traversal security issues.  See, e.g.:
+            // CVE-2001-1267, CVE-2002-0399, CVE-2005-1918, CVE-2007-4131
+            if contains_dot_dot {
+                continue;
+            }
+            // Skip cases where only slashes were seen, because this is
+            // effectively an empty filename.
+            if !seen_non_empty_part {
```

Comment:
I think that the only security issue from not skipping over essentially empty paths is that the permissions of the directory into which the tarball is extracted could be modified.

Note that in my latest patch, I also skip over '.'.